### PR TITLE
Addressing Memory management of Molecules

### PR DESCRIPTION
**Pull request by JoshuaSBrown**
Part 7

### CODE REVIEW

File: include/votca/csg/topology.h [hunk 1]
```diff
@@ -447,9 +447,8 @@ inline Bead *Topology::CreateBead(Bead::Symmetry symmetry, std::string name,
 }
 
 inline Molecule *Topology::CreateMolecule(std::string name) {
-  Molecule *mol = new Molecule(_molecules.size(), name);
-  _molecules.push_back(mol);
-  return mol;
+  _molecules.emplace_back(new Molecule(_molecules.size(), name));
```
Comment (JensWehner) [inline include/votca/csg/topology.h]: I think here just using `(_molecules.size(), name)` could be enough, because emplace_back forwards the things to the constructor. Otherwise use `make_unique` normally no `new` should appear in the code

File: src/libcsg/modules/io/lammpsdatareader.cc [hunk 2]
```diff
@@ -88,7 +88,7 @@ bool LAMMPSDataReader::ReadTopology(string file, Topology &top) {
   NextFrame(top);
 
   fl_.close();
-  for (Molecule *mol : top.Molecules()) {
+  for (auto &mol : top.Molecules()) {
```
Comment (JensWehner) [inline src/libcsg/modules/io/lammpsdatareader.cc:91]: why this change? `auto`does not add anything for me.

File: src/libcsg/topology.cc [hunk 3]
```diff
@@ -245,7 +240,7 @@ void Topology::SetBeadTypeMass(string name, double value) {
 void Topology::CheckMoleculeNaming(void) {
   map<string, Index> nbeads;
 
-  for (Molecule *mol : _molecules) {
+  for (const auto &mol : _molecules) {
```
Comment (JensWehner) [inline src/libcsg/topology.cc:235]: like the `const` do not get the `auto`

**Review by JensWehner — APPROVED** [review]
It is an improvement although I do not like using auto that much. If @junghans likes it I am okay.

File: src/libcsg/modules/io/lammpsdatareader.cc (hunk 2 above)
Comment (JoshuaSBrown) [reply src/libcsg/modules/io/lammpsdatareader.cc:91]: Ya, I've begun to swing towards using auto more, specifically in cases where the type is really long and 2 if I were to change from using a unique_ptr to a smart pointer I would have to update each and every one of these for loops. It is way more work even if it is less explicit, and there are only a few of us.

File: include/votca/csg/topology.h (hunk 1 above)
Comment (JoshuaSBrown) [reply include/votca/csg/topology.h]: I cannot get either of your suggestions to compile. The first one is because _molecules now stores unique pointers so it is expecting the argument to create a unique pointer. The second one becomes really long because I will also have to call move.

Comment (JensWehner) [reply include/votca/csg/topology.h]:
```suggestion
  _molecules.emplace_back(std::make_unique<Molecule>(_molecules.size(), name));
```

**Comment by junghans** [comment]
```
/__w/csg/csg/votca/xtp/src/libxtp/md2qmengine.cc:188:49: error: cannot convert 'const std::unique_ptr<votca::csg::Molecule>' to 'const votca::csg::Molecule*' in initialization
  188 |   for (const csg::Molecule* mol : top.Molecules()) {
      |                                                 ^
gmake[2]: *** [xtp/src/libxtp/CMakeFiles/votca_xtp.dir/build.make:1125: xtp/src/libxtp/CMakeFiles/votca_xtp.dir/md2qmengine.cc.o] Error 1
```

**Comment by junghans** [comment]
> It is an improvement although I do not like using auto that much. If @junghans likes it I am okay.

Ok with me.

**Review by JensWehner — APPROVED** [review]

**Comment by junghans** [comment]
@JensWehner 
```
In file included from /usr/include/c++/10/bits/locale_conv.h:41,
                 from /usr/include/c++/10/locale:43,
                 from /usr/include/boost/lexical_cast/detail/converter_lexical_streams.hpp:43,
                 from /usr/include/boost/lexical_cast/detail/converter_lexical.hpp:54,
                 from /usr/include/boost/lexical_cast/try_lexical_convert.hpp:44,
                 from /usr/include/boost/lexical_cast.hpp:32,
                 from /__w/votca/votca/tools/include/votca/tools/tokenizer.h:26,
                 from /__w/votca/votca/csg/src/libcsg/beadlist.cc:19:
/usr/include/c++/10/bits/unique_ptr.h: In instantiation of 'typename std::_MakeUniq<_Tp>::__single_object std::make_unique(_Args&& ...) [with _Tp = votca::csg::Molecule; _Args = {long unsigned int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&}; typename std::_MakeUniq<_Tp>::__single_object = std::unique_ptr<votca::csg::Molecule>]':
/__w/votca/votca/csg/include/votca/csg/topology.h:447:77:   required from here
/usr/include/c++/10/bits/unique_ptr.h:962:30: error: 'votca::csg::Molecule::Molecule(votca::Index, std::string)' is private within this context
  962 |     { return unique_ptr<_Tp>(new _Tp(std::forward<_Args>(__args)...)); }
      |                              ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /__w/votca/votca/csg/include/votca/csg/topology.h:37,
                 from /__w/votca/votca/csg/src/libcsg/beadlist.cc:23:
/__w/votca/votca/csg/include/votca/csg/molecule.h:92:3: note: declared private here
   92 |   Molecule(Index id, std::string name) : _id(id), _name(name) {}
      |   ^~~~~~~~
gmake[2]: *** [csg/src/libcsg/CMakeFiles/votca_csg.dir/build.make:66: csg/src/libcsg/CMakeFiles/votca_csg.dir/beadlist.cc.o] Error 1
```

**Comment by junghans** [comment]
@JoshuaSBrown ping

**Comment by junghans** [comment]
This will need a change in xtp as well:
```
[ 67%] Building CXX object xtp/src/libxtp/CMakeFiles/votca_xtp.dir/molden.cc.o
cd /__w/csg/csg/builddir/xtp/src/libxtp && /usr/lib64/ccache/g++ -Dvotca_xtp_EXPORTS -I/__w/csg/csg/builddir/xtp/src/libxtp -I/__w/csg/csg/votca/xtp/src/libxtp -I/__w/csg/csg/votca/xtp/include -I/__w/csg/csg/builddir/xtp/include -I/__w/csg/csg/builddir/xtp/include/votca/xtp -I/__w/csg/csg/votca/csg/include -I/__w/csg/csg/votca/tools/include -I/__w/csg/csg/builddir/tools/include -I/__w/csg/csg/builddir/tools/include/votca/tools -isystem /usr/include/eigen3 -isystem /usr/include/libint2 -isystem /usr/include/libecpint -Wall -Wextra -Wpedantic -Wshadow -Wconversion -Werror -Wno-deprecated-copy -O2 -DNDEBUG -fPIC -fopenmp -std=c++14 -o CMakeFiles/votca_xtp.dir/molden.cc.o -c /__w/csg/csg/votca/xtp/src/libxtp/molden.cc
/__w/csg/csg/votca/xtp/src/libxtp/md2qmengine.cc: In member function 'votca::xtp::Topology votca::xtp::Md2QmEngine::map(const votca::csg::Topology&) const':
/__w/csg/csg/votca/xtp/src/libxtp/md2qmengine.cc:188:49: error: cannot convert 'const votca::csg::Molecule' to 'const votca::csg::Molecule*' in initialization
  188 |   for (const csg::Molecule* mol : top.Molecules()) {
      |                                                 ^
```

**Comment by junghans** [comment]
@votca-bot format
